### DISCUSSION TITLE
Prisma - multi tenancy - shared database strategy

### DISCUSSION THREAD

**Discussion by shauliz** [prisma/prisma#11601] [Q&A]
Hi all,
I'm new to Prisma and love it :) (used to work with Sequelize).
I wanted to add multi-tenancy support (mainly for SaaS products) using a shared database strategy, meaning, one database and on each table, a tenant ID column. The idea is that from a developer point of view, there is no need to worry about adding a where clause to each CRUD operation, we will do it for him/her.

As I see it, there were 3 challenges:

1. Extracting the tenant id from a request/message - I've added an express middleware for that.
2. Storing the tenant id in a way switching context in an async operations won't affect it - solve it with async local storage.
3. Making sure each database query is with a `where` clause and with the proper tenant ID - I've added a Prisma middleware for that.

I've wrote (me and @yakirgot) a new npm package for all of it [here](https://www.npmjs.com/package/node-express-multitenant), and the [code is here.](https://github.com/shauliz/node-express-multitenant).

I have two questions:
1. in different languages, I used to work with database hooks or scope to get this functionality and I wonder if middleware is the right approach in prisma. Am I missing anything?
2. My `where` clause validation is pretty simple and straightforward, I would love to get your opinion on it.

PS, you are all more than welcome to contribute, there is still much work needed :)

Thanks in advance,
Shaul

**Comment by shulmaneyal** [comment]
Very cool work @shauliz! Will try to integrate it to our project.

One thing that caused me an issue when trying to use a middleware for scoping by tenant is that the tenantId had to be declared as nullable in the prisma schema. As we're using prisma migrate, my understanding is that then you either have a nullable column in the database schema or have to clean the generated migrations code on every new migration.

**Reply by shauliz** [reply]
Hi Eyal, that's great to hear!
I don't understand why it has to be nullable? Do you plan to have records without tenant it?
If so, you can use default value as -1 and make sure to use the `ignoremultitenancy` flag when you want to query it.

**Comment by shulmaneyal** [comment]
It's just that when creating an object with a non-nullable property in prisma AFAIK you need to provide a value for the property even if a later middleware enriches it.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
